src/wp-includes/blocks.php

```diff
@@ -1166,12 +1166,33 @@ function parse_blocks( $content ) {
  */
 function do_blocks( $content ) {
 	$blocks = parse_blocks( $content );
+
+	/**
+	 * Filter that allows plugins to inspect or manipulate the array of parsed
+	 * block objects. The parsed blocks form a hierarchy, rather than a flat
+	 * array, as blocks can be nested within each other during parsing.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param array[] $blocks Array of parsed block objects.
+	 */
+	$blocks = apply_filters( 'do_blocks_pre_render', $blocks );
+
 	$output = '';
 
 	foreach ( $blocks as $block ) {
 		$output .= render_block( $block );
 	}
 
+	/**
+	 * Filter to allow plugins to inspect or manipulate the final HTML output.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param string $output String of rendered HTML.
+	 */
+	$output = apply_filters( 'do_blocks_post_render', $output );
+
 	// If there are blocks in this content, we shouldn't run wpautop() on it later.
```

Comment:
The interesting part of introducing these filters is that we could even consider moving the logic below as a separate filter applied with `do_blocks_post_render`.

Reply:
`do_blocks_post_render` only gets `$output` (HTML), but not `$content` (parsed blocks).

So can that logic be applied using the `do_blocks_pre_render` filter, or maybe it makes sense to pass `$content` also to the `do_blocks_post_render` filter?

Reply:
> do_blocks_post_render only gets $output (HTML), but not $content (parsed blocks).

I missed that.

In https://github.com/WordPress/wordpress-develop/pull/5023#pullrequestreview-1582962695 @dmsnell mentioned:

>  I wondered if we would want to pass additional args: the original HTML content, the post id, etc

That might make sense to pass the `$content` with both hooks. I often see additional parameters passed with the hook that help with processing in plugins.

`$post_id` isn't explicitly used in the function, and you can always read it with the `get_post` call.

Reply:
> That might make sense to pass the $content with both hooks.

Ok, let's see. Right now the filters are:

- `$blocks = apply_filters( 'do_blocks_pre_render', $blocks );`
- `$output = apply_filters( 'do_blocks_post_render', $output );`

And we have these variables:

- `$content`: Post content (unprocessed).
- `$blocks`: Array of parsed blocks.
- `$output`: Processed post content

It makes sense to pass `$content` to `pre`, but which variables do you think we should pass to `post`? Only `$content`?

Reply:
Again, passing `$blocks` may depend on [how fast is `parse_blocks( $content )`](https://github.com/WordPress/wordpress-develop/pull/5023#issuecomment-1683610594) and if it makes sense to try to avoid doing it more than once.

Reply:
> are you suggesting that we re-parse the output and pass it along in the filter? 

On the contrary. I was simply mentioning that if we do not pass `$blocks` to `do_block_post_render`, plugins that need it can access them by re-parsing, but that means that parsing would be done two (or more) times. And I don't know if that is concerning in terms of performance or not.

> Do we have a compelling use-case we can think of to pass $blocks to the do_blocks_post_render function?

Not that I can think of.

---

Ok, so what about this?

- `$blocks = apply_filters( 'do_blocks_pre_render', $blocks, $content );`
- `$output = apply_filters( 'do_blocks_post_render', $output, $content );`

The use case for `do_blocks_post_render` is to be able to move [this custom logic](https://github.com/WordPress/wordpress-develop/blob/5abfc9c4a2c81b82bdb489211db2493517e5b26b/src/wp-includes/blocks.php#L1196-L1201) outside of the `do_blocks` function.

I don't know any use case for passing `$content` to `do_blocks_pre_render`, but if we're passing it to the `post` filter, I think it makes sense to me to pass it to the `pre` filter as well. Anyway, if you prefer, we can leave omit that one until there's a clear use case.

cc: @gziolo @dmsnell

Reply:
My vote is almost always going to fall on the side of waiting until we have a need to provide a solution, which in this case is for omitting the second parameter. I'm not sure how someone is supposed to associate `$content` to `$blocks` or what the difference between `$content` and `$output` is, or again, how to associate them.

That's just my vote. As noted before though, the easier option is leaving them out and adding them once we need them, since they become additional arguments. It's definitely harder to remove them if we realize we shouldn't expose them.

Reply:
Fair enough, let's not add any additional arguments from the beginning then.